src/main/scala/mesosphere/marathon/core/task/termination/impl/TaskKillProgressActor.scala

```diff
@@ -0,0 +1,70 @@
+package mesosphere.marathon.core.task.termination.impl
+
+import akka.actor.{ Actor, ActorLogging, Props }
+import mesosphere.marathon.KillingTasksFailedException
+import mesosphere.marathon.core.task.Task
+import mesosphere.marathon.core.event.MesosStatusUpdateEvent
+
+import scala.concurrent.Promise
+import scala.collection.mutable
+import scala.util.Try
+
+/**
+  * This actor watches over a given set of taskIds and completes a promise when
+  * all tasks have been reported terminal (including LOST et al)
+  *
+  * @param taskIds the taskIds that shall be watched.
+  * @param promise the promise that shall be completed when all tasks have been
+  *                reported terminal.
+  */
+class TaskKillProgressActor(taskIds: mutable.HashSet[Task.Id], promise: Promise[Unit]) extends Actor with ActorLogging {
+  // TODO: if one of the watched task is reported terminal before this actor subscribed to the event bus,
+  //       it won't receive that event. should we reconcile tasks after a certain amount of time?
+
+  // this should be used for logging to prevent polluting the logs
+  val name = "TaskKillProgressActor" + self.hashCode()
+
+  override def preStart(): Unit = {
+    if (taskIds.nonEmpty) {
+      context.system.eventStream.subscribe(self, classOf[MesosStatusUpdateEvent])
+      log.info("Starting {} to track kill progress of {} tasks", name, taskIds.size)
+    } else {
+      promise.tryComplete(Try(()))
+      log.info("premature aborting of {} - no tasks to watch for", name)
+      context.stop(self)
+    }
+  }
+
+  override def postStop(): Unit = {
+    context.system.eventStream.unsubscribe(self)
+
+    if (!promise.isCompleted) {
+      val msg = s"$name was stopped before all tasks are killed. Outstanding: ${taskIds.mkString(",")}"
+      log.error(msg)
+      promise.failure(new KillingTasksFailedException(msg))
+    }
+  }
+
+  override def receive: Receive = {
+    case Terminal(event) if taskIds.contains(event.taskId) =>
+      log.debug("Received terminal update for {}", event.taskId)
+      taskIds.remove(event.taskId)
+      if (taskIds.isEmpty) {
+        log.info("All tasks watched by {} are killed, completing promise", name)
+        val success = promise.tryComplete(Try(()))
+        if (!success) log.error("Promise has already been completed in {}", name)
+        context.stop(self)
+      } else {
+        log.info("{} still waiting for {} tasks to be killed", name, taskIds.size)
+      }
+
+    case _ => // ignore
```

Comment:
ha, I usually have a `log.debug` in there while during implementation and tend to miss removing it .. removed.